SPEC.md

```diff
@@ -442,17 +442,19 @@ Note that for the purpose of new lines, CRLF is considered _a single newline_.
 ## Full Grammar
 
 ```
-nodes := linespace* (node nodes?)? linespace*
+nodes := (line-space* node)* line-space*
 
-node := ('/-' node-space*)? type? identifier (node-space+ node-prop-or-arg)* (node-space* node-children ws*)? node-space* node-terminator
-node-prop-or-arg := ('/-' node-space*)? (prop | value)
-node-children := ('/-' node-space*)? '{' nodes '}'
-node-space := ws* escline ws* | ws+
+line-space := newline | ws | single-line-comment | '/-' node-space* node
+node-space := ws* escline ws* | ws+ | '/-' node-space* (node-prop-or-arg | node-children)
```

Comment:
Given that we're treating this as breaking anyway, I want to propose writing this instead as

```suggestion
node-space := plain-node-space+ ('/-' plain-node-space* (node-prop-or-arg | node-children) plain-node-space+)
plain-node-space := ws* escline ws* | ws+
```

in order to require separating whitespace around slashdashes (even though that'll make my implementation more annoying, I think).

This split is actually necessary to avoid allowing `/- /-{} {}`.... which kdl-rs also allows 🙃

Reply:
As written, this does still allow `node /-"val"{}` (which is `node {}`), as space is not required before a children block. I'm still slightly biased towards requiring space before children blocks or never requiring space (for uniformity), but I see the appeal of allowing omitting the space before the children block. An interesting approach, if somewhat involved, is to allow `node{}` but not `node "val"{}`:

```
node := type? identifier node-entries node-space* node-terminator
node-entries := node-children | (node-space+ (prop | value))* (node-space+ node-children)?
```

I kinda like this, but this is drifting further into more involved changes.

Reply:
My inclination with all this is to ask "which one is the simplest to implement that will work as expected in most cases?" One of KDL's design goals is to be relatively straightforward to implement, even if it results in some "funny-looking" documents.

Reply:
In that case, I can say that from my experience:

- With the grammar changed to just classify slashdash as whitespace, implementing is fairly simple either way;
- Requiring whitespace only sometimes is more annoying than being consistent (you have to remember if you ate whitespace to decide later if there's an error); and
- Not requiring whitespace is simpler to implement than requiring whitespace, as less code is easier (it's quite literally just delete the code which requires the whitespace). Additionally, never requiring whitespace means that it's possible to use a lexer which discards the trivia (whitespace, comments) eagerly.

The closer the grammar can be to satisfying being a PEG the easier it'll be to implement, as PEG is quite literally just a declarative recursive descent parser.

Reply:
Shouldn't this then also change?

```diff
-line-space := newline | ws | single-line-comment | '/-' node-space* node
+line-space := newline | ws | single-line-comment | '/-' plain-node-space* node
```

Otherwise the following would be valid and imo pretty confusing

```kdl
/- /-"commented" /-{ commented; } commented "commented" { commented; }
```

Reply:
Ah yes, I agree, that's a good change as well, if we're distinguishing plain whitespace.